DOCSP-13833 crud cursor (#79)
mongodb/docs-golang: source/fundamentals/crud/read-operations/cursor.txt

Each section uses the following ``cursor`` variable, which is a
``Cursor`` struct that contains all the documents in a collection:

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin find
   :end-before: end find

.. important::

   A cursor is not goroutine safe. Do not use the same cursor in
   multiple goroutines at the same time.

.. _individual-documents-golang:

Retrieve Documents Individually
-------------------------------

To retrieve documents from your cursor individually while blocking the
current goroutine, use the ``Next()`` method.

The method returns a document if each of the following is true:

- A document is currently or will later be available
- No errors occurred
- The context didn't expire

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin cursor next
   :end-before: end cursor next

Tailable Cursor
~~~~~~~~~~~~~~~

To attempt retrieving a document from a :manual:`tailable cursor
</core/tailable-cursors/>`, use the ``TryNext()`` method.

The method returns a document if each of the following is true:

> - A document is currently or will later be available

Reviewer: Optional: `TryNext()` doesn't wait for a document to be available before returning. Update to
> A document is currently available

Sample Cursor
~~~~~~~~~~~~~

Each section uses the following ``cursor`` variable, which is a
``Cursor`` struct that contains all the documents in a collection:

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin find
   :end-before: end find

.. important::

   A cursor is not goroutine safe. Do not use the same cursor in
   multiple goroutines at the same time.

.. _individual-documents-golang:

Retrieve Documents Individually
-------------------------------

To retrieve documents from your cursor individually, use the
``Next()`` method.

This method blocks the current goroutine. That means the function
returns a document if each of the following is true:

- A document is currently or will later be available
- No errors occurred
- The ``MongoClient`` didn't exhaust the cursor
- The context didn't expire

.. include:: /includes/fundamentals/exhausted-cursor-definition.rst

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin cursor next
   :end-before: end cursor next

Documents Only in the Cursor

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: [Responding to this comment thread](https://github.com/mongodb/docs-golang/pull/79#discussion_r746946026) as I it seems like I can't respond through the github UI anymore.

The [tailable cursors](https://docs.mongodb.com/manual/core/tailable-cursors/) page says the following: 

> However, for capped collections you may use a Tailable Cursor that remains open after the client exhausts the results in the initial cursor.

I agree this wording is unclear as to whether tailable cursors are only available in capped collections. Could a ticket be filed with server team to clarify whether or not cursors are only available for capped collections?

The Go [API docs say](https://pkg.go.dev/go.mongodb.org/mongo-driver@v1.7.2/mongo#Cursor.TryNext):

> This is only recommended for use with tailable cursors as a non-blocking alternative to Next

Could this section be titled to better reflect this use case? I'm not sure if this is an equivalent condition as "Documents only being in the Cursor". Could the title be changed to reflect "blocking" vs "non-blocking"?

> To attempt retrieving a document from your cursor, use the ``TryNext()``
> method.

Reviewer 2: Suggestion: I'm not sure "attempt retrieving" has a precise meaning. Could this be changed to reflect more closely the actions the method is taking?

```suggestion
To retrieve a document from your cursor if the document is currently in memory, use the ``TryNext()``
method.
```

This method is unable to block the current goroutine. That means the

> function returns a document if each of the following is true:

Reviewer 2: Suggestion: Similar to previous comment, I think "does not block" is separate from the conditions under which the method returns a document.

- A document is currently available
- No errors occurred
- The ``MongoClient`` didn't exhaust the cursor
- The context didn't expire

> .. include:: /includes/fundamentals/exhausted-cursor-definition.rst

Reviewer 2: Suggestion: As admonitions are less likely to be read by readers, could this definition be removed from an admonition and placed in the main text?

Suggestion: Could this information only be presented 1 time on the page? I believe this can be defined once in the "Retrieve Documents Individually" section as it is the parent section to all sections that contain this admonition.

Author: After discussing this section with NL, we decided to remove the "exhaust the cursor" bullet since by definition, thats the first bullet (where it says IF a document exists). I'll also be removing this note since there is no need to define it anymore.

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin cursor try next
   :end-before: end cursor try next

Retrieve All Documents
----------------------

To populate an array with all of your query results, use the ``All()``
method:

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin cursor all
   :end-before: end cursor all

.. important:: Memory

   If the number and size of documents returned by your query exceeds
   available application memory, your program will crash. If you except

>    a large result set, you should :ref:`consume the cursor iteratively

Reviewer 2: Nit: I think this can be made 2nd person

```suggestion
   a large result set, you should :ref:`consume your cursor iteratively
```

   <individual-documents-golang>`.

Close the Cursor
----------------

When your application no longer needs to use a cursor, close the cursor

> with the ``Close()`` method. This method frees up your cursor's
> consumption of resources in both the client application and the 

Reviewer 2: Suggestion: I think "frees up" is idiomatic. Consider deleting "up".
Suggestion: I think "frees consumption..." should be "frees the resources" (I do not think "consumption" can be freed). 

```suggestion
with the ``Close()`` method. This method frees the resources your cursor consumes in both the client application and the 
```

MongoDB server.

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin close
   :end-before: end close

.. note:: 

>    Close the cursor when you :ref:`retrieve documents individually
>    <individual-documents-golang>` because those functions make a cursor
>    :manual:`tailable </core/tailable-cursors/>`.

Reviewer 2: Suggestion: Could this either explain why a cursor being "tailable" requires that you must close the cursor or remove the word "tailable"?

Possible Fix: Maybe a link to "tailable" could be presented in the "Retrieve Documents Individually" section where tailable cursors are created?

```suggestion
   When you :ref:`retrieve documents individually
   <individual-documents-golang>`, you must close the cursor.
```

.. _cursor-golang:

=========================
Access Data From a Cursor
=========================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to access data with a **cursor**.

A cursor is a mechanism that allows an application to iterate over
database results while holding only a subset of them in memory at a
given time. Read operations that match multiple documents use a cursor
to return those documents in batches as opposed to all at once.

Sample Cursor
~~~~~~~~~~~~~

The following sections use a ``cursor`` variable, which is a cursor that

> retrieved all the documents in a collection with the ``Find()`` function.

Author: I agree with the first two and half of the third point-- For the third, I think it's ok to omit "returned from a find operation" because the snippet shows the `Find()` call.

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin find
   :end-before: end find

.. important::

   A cursor is not goroutine safe. Do not use the same cursor in
   multiple goroutines at the same time.

Retrieve All Documents
----------------------

To populate an array with all of your query results, use the ``All()``
method:

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin cursor all
   :end-before: end cursor all

.. important:: Memory

   This method crashes your application if your query retrieves a number
   of documents larger than the size of your array. If this occurs,

>    :ref:`retrieve the documents individually <individual-documents-golang>`.

Reviewer 3: I think it would be better to just state what this does.

If the number of size of documents returned by your query exceeds available application memory, your program will crash. If you except a large result set, you should consume the cursor iteratively <...>.

.. _individual-documents-golang:

Retrieve Documents Individually
-------------------------------

To retrieve documents from your cursor individually, use the
``Next()`` method.

This method blocks the current goroutine. That means the function
returns a document if each of the following is true:

- A document is currently or will later be available
- No errors occurred
- The ``MongoClient`` didn't exhaust the cursor

> - The context didn't expire

Reviewer 2: I believe "block" has more to do with running on the same thread/goroutine or running asynchronously. So I think "blocking" is more akin to "executes on the same goroutine" rather than "will return a document given these conditions".

Could this be rephrased to avoid suggesting that "blocks the current goroutine" is defined as the conditions under which `Next()` does not throw an error?

.. tip::

   Iterating through your documents individually removes the possibility
   of crashing your application and allows you to :ref:`access a cursors
   elements <cursor-elements-golang>`.

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin cursor next
   :end-before: end cursor next

> .. note::
>    
>    If you previously used the cursor to iterate some documents, the
>    driver excludes those documents when you iterate again. 

Reviewer 2: Suggestion: Could "previously used the cursor to iterate" be clarified? I'm not sure "iterate some documents" has a precise meaning.
Suggestion: What is meant by "iterate again"? Is this a new cursor instance, or the same cursor instance? I believe you can iterate over an identical result set if you remake the cursor?
Suggestion: Can a recommendation be provided if a reader wants to process their documents first to last or last to first in the same result set?

```suggestion
.. note:: One Way
   
   If you use the ``Next()`` method to iterate over your result set, you can only process your documents from first 
   document to last document. You cannot process your documents in reverse order. If you need to process your
   documents in both first to last and last to first order, consider 
   :ref:`loading all of your documents into a slice <LINK TO RETRIEVE ALL DOCUMENTS SUBSECTION>`. 
```

Author: Discussed this with Nathan and agreed it would be better to remove this note since this happens by definition.

> Documents Only in the Cursor
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> To attempt retrieving a document from your cursor, use the ``TryNext()``
> method.
> 
> This method is unable to block the current goroutine. That means the
> function returns a document if each of the following is true:
> 
> - A document is currently available
> - No errors occurred
> - The ``MongoClient`` didn't exhaust the cursor
> - The context didn't expire
> 
> .. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
>    :language: go
>    :dedent:
>    :start-after: begin cursor try next
>    :end-before: end cursor try next
> 
> .. note::
>    
>    If you previously used the cursor to iterate some documents, the
>    driver excludes those documents when you iterate again. 
> 

Reviewer 2: Suggestion: After looking at the API docs for the [`TryNext()` method](https://pkg.go.dev/go.mongodb.org/mongo-driver@v1.7.2/mongo#Cursor.TryNext) it seems like this is only meant to be used for watching for changes in capped collections. Could the section be tweaked to reflect this?

Question: What is the exact mechanism by which a tailable/non-tailable cursor is created? I can't find this information. Are all cursors created by the go driver tailable or only some?

Rough draft :

```suggestion
Wait For Documents After The Cursor Is Exhausted
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you need to wait for more documents after your cursor is exhausted, use a
**tail-able cursor**. A tail-able cursor is a cursor that remains open after it is
exhausted and watches for additional documents that match your query.

.. important:: Only Available for Capped Collections

   Tailable cursors are only available for :manual:`capped collections <LINK TO MANUAL>`.

Use the ``TryNext()``method to create a tail-able cursor:

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin cursor try next
   :end-before: end cursor try next

For more information, see :manual:``Tailable Cursors <manual/core/tailable-cursors/>`` in the MongoDB manual.
```

Author: I don't think this is a capped collection. According to [our docs](https://docs.mongodb.com/manual/core/capped-collections/#tailable-cursor), you CAN use a tailable cursor with one, but not that it will by default. Additionally, [the spec](https://github.com/mongodb/specifications/blob/747b7480c47900827ff2b64f23c2a14636803d1a/source/crud/crud.rst#read) doesn't say anything about a capped collection, which makes me think it isn't or it's irrelevant to its operation.

.. _cursor-elements-golang:

Access Cursor Elements

> ----------------------

Reviewer 2: Suggestion: Can this section be deleted? I believe it largely duplicates the content of the Go API Documentation, and takes up significant real-estate on this page.

As an alternative, could the reader be linked to the [Working with BSON page](https://docs-mongodbcom-staging.corp.mongodb.com/golang/docsworker-xlarge/DOCSP-13833-CRUDCursor/fundamentals/bson/#unmarshalling) page earlier in the document when the `Decode()` method is used in the `Next()` example? 

I think `Current` and the "Other Elements" can be left to the API documentation, as why a reader might be interested in them is not discussed here.

You can access a cursors elements before it's **exhausted**. A cursor is
exhausted when the ``MongoClient`` iterates through its results and
reaches its last element. Afterwards, the cursor won't respond to

> functions to access its elements.

Reviewer 2: Issue: Can this definition be moved out of this section to somewhere earlier on the page? I believe this word is used before it is defined.

   :end-before: end current

Go Value Document
~~~~~~~~~~~~~~~~~

To access the current document as a Go value, use the ``Decode()``
function. 

This function unmarshals the current document into the :ref:`BSON
interface <bson-types>` you specify. It returns any errors from the
unmarshalling process without any modification.

.. code-block:: go

   var result bson.D
   if err := cursor.Decode(&result); err != nil {
      log.Fatal(err)
   }
   fmt.Println(result)

Other Elements
~~~~~~~~~~~~~~

You can also retrieve the following elements:

- `The number of documents remaining in the batch. <{+api+}/mongo#Cursor.RemainingBatchLength>`__
- `The cursor ID. <{+api+}/mongo#Cursor.ID>`__ 
- `The cursor errors encountered. <{+api+}/mongo#Cursor.Err>`__

.. important:: Retrieving the Cursor ID

   If you want to retrieve the cursor ID, include the option to set your
   batch size. The batch size needs to be a number less than or equal to
   the number of documents you expect to retrieve.
   
   .. code-block:: go
      
      opts := options.Find().SetBatchSize(<a number>)
      cursor, err := coll.Find(context.TODO(), bson.D{}, opts)
      if err != nil {
         panic(err)

>       }

Reviewer 2: Suggestion: This is useful information, but I think should be omitted from this page as a discussion of why one would require the cursor ID is not included (I can't think of a use case off the top of my head for why one would need this outside of the Driver's source code, but maybe there is a use case).

Could a ticket be made to include this information in the API docs for the Driver? Could this note be removed from the page?

   If you omit this option, set the batch size to a number greater than
   the elements retrieved, or the ``MongoClient`` exhausts the cursor,
   the cursor ID is ``0``.

Close the Cursor
----------------

When your application no longer needs to use a cursor, close the cursor

> with the ``Close()`` method. This method free's up your cursor's

Reviewer 2: Nit: I think this should not have the apostrophe

```suggestion
with the ``Close()`` method. This method frees up your cursor's
```

consumption of resources in both the client application and the 
MongoDB server.

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin close
   :end-before: end close

> .. note:: 
> 
>    Close the cursor when you retrieve documents individually because
>    those functions make a cursor :manual:`tailable
>    </core/tailable-cursors/>`.

Reviewer 2: Question/Suggestion: Are all cursors tailable? How do you open a tail-able/non-tailalbe cursor? The API docs seem to suggest that not all cursors are tailable? Could the mechanism to open a tailable vs non-tailable cursor be confirmed? If the driver contains both tailable and non-tailable cursors, could this note be clarified? 

```suggestion
.. note:: Close Tail-able Cursors 

   As :ref:`tail-able cursors <PRECEDING SECTION>` remain open indefinitely, make sure to manually close 
   tail-able cursors.
```

.. _cursor-golang:

=========================
Access Data From a Cursor
=========================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to access data with a **cursor**.

A cursor is a mechanism that allows an application to iterate over
database results while holding a subset of them in memory at a given
time. The Go driver uses a cursor in some read operations to return
matched documents in batches as opposed to all at once.

> Retrieve Sample Data
> ~~~~~~~~~~~~~~~~~~~~
> 
> Each section assumes you retrieve all the documents in your collection
> with the ``Find()`` function.

Reviewer 2: Suggestion: As the examples on this page do not reference a sample data set, could "Retrieve Sample Data" be changed to reflect the "sample cursor" examples reference?

Rough draft of potential solution:

```suggestion
Example Cursor
~~~~~~~~~~~~

The sections in this guide reference the `cursor` variable from the following code snippet. The cursor variable contains a cursor that contains all documents in a collection.
```

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin find
   :end-before: end find

Access Cursor Elements

> ----------------------

Reviewer 2: Suggestion: As readers must use begin to process their cursor with the content currently located in the "Usage Patterns" section before accessing cursor elements, could this section be moved after the content currently located in the "Usage Patterns" section?

You can access a cursors elements before it's **exhausted**. A cursor is
exhausted when the ``MongoClient`` iterates through its results and
reaches its last element. Afterwards, the cursor won't respond to
functions to access its elements.

.. important::

   A cursor is not goroutine safe. Do not use the same cursor in
   multiple goroutines at the same time.

Current Document
~~~~~~~~~~~~~~~~

When :ref:`iterating through your cursors results
<cursor-iteration-golang>`, you can view each document as a ``BSON`` or
Go value.

BSON

> ````

Reviewer 2: Issue: [Three levels of headings](https://docs.mongodb.com/meta/style-guide/style/titles-and-headings/#style-and-structure)

To access the current document as raw ``BSON``, use the ``Current``
property.

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin current

>    :end-before: end current

Reviewer 2: Suggestion/Question: I'm not sure why this code snippet displays a print statement. Can the print statment be removed if it is not necessary?

Author: It isn't "necessary", but since the Go Value example prints out each result, I think it would be odd to have this one not print as well.

Go Value
````````

To access the current document as a Go value, use the ``Decode()``
function. 

This function unmarshals the current document into the :ref:`BSON
interface <bson-types>` you specify. It returns any errors from the
unmarshalling process without any modification.

.. code-block:: go

   var result bson.D
   if err := cursor.Decode(&result); err != nil {
      log.Fatal(err)
   }
   fmt.Println(result)

Remaining Batch Length
~~~~~~~~~~~~~~~~~~~~~~

To retrieve the amount of documents remaining in the current batch, use
the ``RemainingBatchLength()`` function.

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin remaining batch length
   :end-before: end remaining batch length

.. note::

   If the amount is ``0`` and you call the ``Next()`` or ``TryNext()``

>    function, it performs a network request to fetch the next batch.

Reviewer 2: Suggestion: I do not believe the remaining batch length section [expands on the API documentation](https://pkg.go.dev/go.mongodb.org/mongo-driver@v1.7.2/mongo#Cursor.RemainingBatchLength). Could this content be removed or replaced with a link to the API documentation? This note applies to all subsections of the "Access Cursor Elements" section that do not expand on the the API documentation.

function. The error is ``nil`` if no errors occurred.

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin err
   :end-before: end err

.. _cursor-iteration-golang:

Usage Patterns
--------------

A cursor allows you to access query results one document at a time,
abstracting away network and caching logic. 

.. note::
   
   If you previously used the cursor to iterate some documents, the
   driver excludes those documents when you iterate again. 

Functional Iteration
~~~~~~~~~~~~~~~~~~~~

Functional iteration functions make your cursor non-tailable, which
means the following occurs:

- The cursor automatically closes after the ``MongoClient`` exhausts it.
- The cursor stops retrieving documents even if you insert them.

All
```

To iterate through all results, use the ``All()`` function. 

This function unmarshals all the documents stored in a cursor into an
array at the same time. 

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin cursor all
   :end-before: end cursor all

> Manual Iteration
> ~~~~~~~~~~~~~~~~
> 
> Manual iteration functions make your cursor tailable, which means the
> following occurs:
> 
> - The cursor remains open after the ``MongoClient`` exhausts it.
> - The cursor continues to retrieve documents as you insert them.
> 
> Next
> ````
> 
> To iterate through each result or to purposely wait for the next
> available document, use the ``Next()`` function. 
> 
> .. tip::
> 
>    While going through each result, you can further specify how you want
>    the documents printed though the :ref:`find operations options
>    <retrieve-options>`.
> 
> This function blocks the current goroutine until a document is
> available, an error occurs, or the context expires. It returns ``true``
> if no errors occur and the ``MongoClient`` didn't exhaust the cursor.
> Otherwise, the function returns ``false``.
>    
> .. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
>    :language: go
>    :dedent:
>    :start-after: begin cursor next
>    :end-before: end cursor next
> 
> Attempt Next
> ````````````
> 
> Alternatively, you can use the ``TryNext()`` function.
> 
> This function attempts to retrieve the next document in a cursor. It's
> non-blocking, which means it returns ``false`` if the ``MongoClient``
> exhausts the cursor, an error occurs, the next document isn't available
> or the context expires. Otherwise, the function returns ``true``.
> 
> .. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
>    :language: go
>    :dedent:
>    :start-after: begin cursor try next
>    :end-before: end cursor try next
> 

Reviewer 2: Suggestion: Could "tailable" be defined or omitted?
Suggestion: Could "Manual Iteration" be given a Task-Oriented title?
Issue: Could "iterate through each" (behavior of `All` method) be changed to "gets the next"?
Issue: I believe `Next` is a method
Suggestion: Could a detailed discussion of return values be saved for API documentation?

This suggestion is a rough draft, but I think it proposes potential solutions to some of the items listed above. 

```suggestion
Iterate Through Your Documents Individually
---------------------------------------------

To get the next result document from your cursor, use the ``Next()`` method.
The next method fetches your result document from Mongo
   
.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin cursor next
   :end-before: end cursor next

.. important:: Blocking

   This function blocks the current goroutine until a document is
   available, an error occurs, or the context expires. 

Retrieve the Next Document if it is currently in the Cursor
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you only want to return the next document if it is already in your Cursor,
use the ``TryNext()`` function:

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin cursor try next
   :end-before: end cursor try next

```

Author: For each section, I didn't include a "The following code snippet ..." description because I felt that it would be reiterating what was mentioned right before it.

> Cursor Cleanup
> --------------
> 
> To free up a cursor's consumption of resources in both the client
> application and the MongoDB server, use the ``Close()`` function. 
> 
> .. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
>    :language: go
>    :dedent:
>    :start-after: begin close
>    :end-before: end close

Reviewer 2: Suggestion: Could the section title be made task-oriented?
 

```suggestion
Close Your Cursor
------------------

Once your application no longer needs to use a cursor, close the cursor with the
``Close()`` method. The close method performs the following actions:

- LIST OF ACTIONS

The following code snippet ... : 

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin close
   :end-before: end close
```
